lib/charms/opensearch/v0/opensearch_plugins.py

```diff
@@ -0,0 +1,348 @@
+
+from charms.opensearch.v0.helper_enums import BaseStrEnum
+from charms.opensearch.v0.opensearch_exceptions import OpenSearchError
+from jproperties import Properties
+
+# The unique Charmhub library identifier, never change it
+LIBID = "3b05456c6e304680b4af8e20dae246a2"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+logger = logging.getLogger(__name__)
+
+
+class OpenSearchPluginError(OpenSearchError):
+    """Exception thrown when an opensearch plugin is invalid."""
+
+
+class OpenSearchPluginMissingDepsError(OpenSearchPluginError):
+    """Exception thrown when an opensearch plugin misses installed dependencies."""
+
+    def __init__(self, name: str, missing_dependencies: List[str]):
+        self._deps = missing_dependencies
+        self._name = name
+
+    def __str__(self):
+        """Converts exception to a string."""
+        return f"Failed to install {self._name}, missing dependencies: {self._deps}"
+
+
+class OpenSearchPluginInstallError(OpenSearchPluginError):
+    """Exception thrown when opensearch plugin installation fails."""
+
+    def __init__(self, name: str, msg: str):
+        self._msg = msg
+        self._name = name
+
+    def __str__(self):
+        """Converts exception to a string."""
+        return f"Failed to install plugin {self._name}: {self._msg}"
+
+
+class OpenSearchPluginRemoveError(OpenSearchPluginError):
+    """Exception thrown when opensearch plugin removal fails."""
+
+    def __init__(self, name: str, msg: str):
+        self._msg = msg
+        self._name = name
+
+    def __str__(self):
+        """Converts exception to a string."""
+        return f"Failed to remove plugin {self._name}: {self._msg}"
+
+
+class PluginState(BaseStrEnum):
+    """Enum for the states possible in plugins' lifecycle."""
+
+    MISSING = "missing"
+    AVAILABLE = "available"
+    ENABLED = "enabled"
+    WAITING_FOR_UPGRADE = "waiting-for-upgrade"
+
+
+class OpenSearchPluginConfig:
+    """Represents the default configuration for a plugin."""
+
+    def __init__(
+        self,
+        config_entries: Optional[Dict[str, Any]] = None,
+        secret_entries: Optional[Dict[str, str]] = None,
+    ):
+        self.config_entries = config_entries
+        self.secret_entries = secret_entries
+
+
+class OpenSearchPlugin:
+    """Abstract class describing an OpenSearch plugin."""
+
+    PLUGIN_PROPERTIES = "plugin-descriptor.properties"
+
+    def __init__(self, plugins_path: str, extra_config: Dict[str, Any] = None):
+        """Creates the OpenSearchPlugin object.
+
+        Arguments:
+          plugins_path: str, path to the plugins folder
+          extra_config: dict, contains the data coming from either the config option
+                              or the relation bag
+        """
+        self._plugins_path = plugins_path
```

Comment:
since our charms run in linux I think it's safe to do the following: 
```suggestion
        self._plugins_path = f"{plugins_path}/{self.PLUGIN_PROPERTIES}"
```